### PR TITLE
Add updateLoanToken rpc

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Add updateLoanToken rpc

#### Which issue(s) does this PR fixes?:

Fixes part of #488

### CODE REVIEW

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      name: 'Token3',
+      priceFeedId,
+      mintable: true,
+      interest: new BigNumber(0.01)
+    }, []])
+    await testing.generate(1)
+
+    await testing.rpc.loan.updateLoanToken({
+      token: 'Token3',
+      symbol: 'ABCDEFGHI', // 9 letters
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:114]: I change to 'x'.repeat(9)

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+      name: 'Token3',
+      priceFeedId
+    })
+    await testing.generate(1)
+
+    const data = await testing.container.call('listloantokens', [])
+    const index = Object.keys(data).indexOf(loanTokenId) + 1
+    expect(data[loanTokenId].token[index].symbol).toStrictEqual('ABCDEFGH') // Only remain the first 8 letters
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:122]: I change to `expect(data[loanTokenId].token[index].symbol).toStrictEqual('x'.repeat(9))`

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 3]
```diff
@@ -35,6 +35,28 @@ export class Loan {
...
+  /**
+   * Updates an existing loan token.
+   *
+   * @param {UpdateLoanToken} loanToken
+   * @param {string} loanToken.token // The tokens's symbol, id or creation tx
+   * @param {string} loanToken.symbol // Token's symbol (unique), no longer than 8
+   * @param {string} loanToken.name // Token's name, no longer than 128
+   * @param {string} loanToken.priceFeedId // Txid of oracle feeding the price
+   * @param {boolean} [loanToken.mintable] // Token's 'Mintable' property, default=true
+   * @param {BigNumber} [loanToken.interest] // Interest rate, default=0
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:48]: I change to this

   * @param {string} loanToken.token The tokens's symbol, id or creation tx
   * @param {string} loanToken.symbol Token's symbol (unique), no longer than 8
   * @param {string} loanToken.name Token's name, no longer than 128
   * @param {string} loanToken.priceFeedId Txid of oracle feeding the price
   * @param {boolean} [loanToken.mintable] Token's 'Mintable' property, default=true
   * @param {BigNumber} [loanToken.interest] Interest rate, default=0
   
For mintable and interest, there should not have default value in the javsacript side. Mintable and interest are defaulted to true and 0 in C++ side (Not Javascript)

File: website/docs/jellyfish/api/loan.md [hunk 4]
```diff
@@ -50,3 +50,27 @@ interface LoanSchemeResult {
...
 }
 ```
+
+## updateLoanToken
+
+Updates an existing loan token.
+
+```ts title="client.loan.updateLoanToken()"
+interface loan {
+  updateLoanToken (loanToken: UpdateLoanToken, utxos: UTXO[] = []): Promise<string>
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply website/docs/jellyfish/api/loan.md:60]: For mintable and interest, there is no default value in the javsacript side. Mintable and interest are defaulted to true and 0 in C++ side (Not Javascript)

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+      priceFeedId
+    })
+    await expect(promise).rejects.toThrow('RpcApiError: \'Test LoanUpdateLoanTokenTx execution failed:\ntoken with key \'Token1\' already exists!\', code: -32600, method: updateloantoken')
+  })
+
+  it('should updateLoanToken if name is more than 128 letters', async () => {
+    await testing.rpc.loan.updateLoanToken({
+      token: 'Token1',
+      symbol: 'Token2',
+      name: 'ABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXY', // 129 letters
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:158]: I change to 'x'.repeat(129)

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 6]
```diff
@@ -35,6 +35,28 @@ export class Loan {
...
+   * @param {UTXO[]} [utxos = []] Specific UTXOs to spend
+   * @param {string} utxos.txid Transaction Id
+   * @param {number} utxos.vout Output number
+   * @return {Promise<string>} LoanTokenId, also the txn id for txn created to update loan token
+   */
+  async updateLoanToken (loanToken: UpdateLoanToken, utxos: UTXO[] = []): Promise<string> {
+    const { token, symbol, name, priceFeedId, mintable, interest } = loanToken
+    return await this.client.call('updateloantoken', [
+      token, { symbol, name, priceFeedId, mintable, interest }, utxos
```
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:57]: I think this is a brilliant idea!

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:114]: I change to 'x'.repeat(9)

File: website/docs/jellyfish/api/loan.md [hunk 7]
```diff
@@ -217,6 +217,30 @@ interface UTXO {
 }
 ```
 
+## updateLoanToken
+
+Updates an existing loan token.
+
+```ts title="client.loan.updateLoanToken()"
+interface loan {
```
Comment (monstrobishi) [inline website/docs/jellyfish/api/loan.md:225]:
```suggestion
class Loan {
```

Comment (jingyi2811) [reply website/docs/jellyfish/api/loan.md:225]: We should use `interface` here. See other functions in the other md files.

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 8]
```diff
@@ -151,6 +151,30 @@ export class Loan {
...
+  /**
+   * Updates an existing loan token.
+   *
+   * @param {UpdateLoanToken} loanToken
+   * @param {string} loanToken.token Previous tokens's symbol, id or creation tx (unique)
+   * @param {string} loanToken.symbol New token's symbol (unique), no longer than 8
+   * @param {string} [loanToken.name] Token's name, no longer than 128
+   * @param {string} loanToken.priceFeedId Txid of oracle feeding the price
+   * @param {boolean} [loanToken.mintable = true] Token's 'Mintable' property
+   * @param {BigNumber} [loanToken.interest = 0] Interest rate
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:163]: all fields are optional.

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 9]
```diff
@@ -151,6 +151,30 @@ export class Loan {
...
+   * @param {UTXO[]} [utxos = []] Specific UTXOs to spend
+   * @param {string} utxos.txid Transaction Id
+   * @param {number} utxos.vout Output number
+   * @return {Promise<string>} LoanTokenId, also the txn id for txn created to update loan token
+   */
+  async updateLoanToken (loanToken: UpdateLoanToken, utxos: UTXO[] = []): Promise<string> {
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:169]: if you can show the separation between old/current `token` parameter and the new properties of the Loan token, it would be better.
```suggestion
  async updateLoanToken (loanTokenIdentifier: string, newLoanTokenDetails: UpdateLoanTokenDetails, utxos: UTXO[] = []): Promise<string> {
```

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 10]
```diff
@@ -151,6 +151,30 @@ export class Loan {
...
+    const defaultData = {
+      mintable: true,
+      interest: 0
+    }
+    const { token, ...tokenData } = loanToken
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:174]: `tokenData` ??

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 9 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:169]: I shall replace with this
`async updateLoanToken (oldToken: string, updateLoanToken: UpdateLoanToken, utxos: UTXO[] = []): Promise<string>`

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 8 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:163]: Why do you say so?

Comment (surangap) [reply packages/jellyfish-api-core/src/category/loan.ts:163]: it's update loan token. so you can basically give any combination of fields that needs to be updated.
Ref- https://github.com/DeFiCh/ain/blob/4492dc3d7a3debd368774e3e2707eb9e208a6a6b/src/masternodes/rpc_loan.cpp#L375
Also, you can check how the CPP side RPC is written there.

File: packages/jellyfish-api-core/src/category/loan.ts (hunk 9 above)
Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts]: I changed to this

`async updateLoanToken (oldToken: string, newTokenDetails: UpdateLoanToken, utxos: UTXO[] = []): Promise<string> {`

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+  })
+
+  it('should updateLoanToken by given symbol', async () => {
+    await testing.rpc.loan.updateLoanToken('Token1', { symbol: 'Token2' })
+    await testing.generate(1)
+
+    const data = await testing.container.call('listloantokens', [])
+    const index = Object.keys(data).indexOf(loanTokenId) + 1
+    expect(data[loanTokenId].token[index].symbol).toStrictEqual('Token2')
+  })
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:142]: duplicate with the test starting from LN:59 ?

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+  it('should not updateLoanToken if symbol is an empty string', async () => {
+    const promise = testing.rpc.loan.updateLoanToken('Token1', { symbol: '' })
+    await expect(promise).rejects.toThrow('RpcApiError: \'Test LoanUpdateLoanTokenTx execution failed:\ntoken symbol should be non-empty and starts with a letter\', code: -32600, method: updateloantoken')
+  })
+
+  it('should not updateLoanToken if token with same symbol was created before', async () => {
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:149]:
```suggestion
  it('should not updateLoanToken if the symbol is used in other token', async () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+  })
+
+  it('should updateLoanToken if mintable is false', async () => {
+    await testing.rpc.loan.updateLoanToken('Token1', { mintable: false })
+    await testing.generate(1)
+
+    const data = await testing.container.call('listloantokens', [])
+    const index = Object.keys(data).indexOf(loanTokenId) + 1
+    expect(data[loanTokenId].token[index].mintable).toStrictEqual(false)
+  })
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:241]: probably can make a single test with name `should updateLoanToken mintable`
then first you can set it to `false` since it is already in `true`. then make it to `true` again.

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+
+    await testing.rpc.loan.updateLoanToken('Token5', { name: 'Token1' }) // Same name as Token1's name
+    await testing.generate(1)
+
+    const data = await testing.container.call('listloantokens', [])
+    const index = Object.keys(data).indexOf(loanTokenId) + 1
+    expect(data[loanTokenId].token[index].name).toStrictEqual('Token1')
+  })
+
+  it('should updateLoanToken with given priceFeedId', async () => {
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:206]: add a test to update price feed id to empty `''` string

**Review by surangap — CHANGES_REQUESTED** [review]
Requesting refactors related to oracle aggregated price changes.
Ref-
https://github.com/DeFiCh/ain/blob/b691ab0e9dc6ae38644c51c2940834c3094e1112/src/masternodes/rpc_loan.cpp#L472

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 15]
```diff
@@ -151,6 +151,25 @@ export class Loan {
...
+   * @param {string} [newTokenDetails.name] Token's name, no longer than 128
+   * @param {string} [newTokenDetails.priceFeedId] Txid of oracle feeding the price
+   * @param {boolean} [newTokenDetails.mintable] Token's 'Mintable' property
+   * @param {BigNumber} [newTokenDetails.interest] Interest rate
+   * @param {UTXO[]} [utxos = []] Specific UTXOs to spend
+   * @param {string} utxos.txid Transaction Id
+   * @param {number} utxos.vout Output number
+   * @return {Promise<string>} LoanTokenId, also the txn id for txn created to update loan token
+   */
+  async updateLoanToken (oldToken: string, newTokenDetails: UpdateLoanToken, utxos: UTXO[] = []): Promise<string> {
```
Comment (canonbrother) [inline packages/jellyfish-api-core/src/category/loan.ts:169]: shouldn't we use `options` at 3rd param??

Comment (jingyi2811) [reply packages/jellyfish-api-core/src/category/loan.ts:169]: https://github.com/DeFiCh/jellyfish/pull/520#discussion_r692697584

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 16]
```diff
@@ -151,6 +151,25 @@ export class Loan {
...
   }
 
+  /**
+   * Updates an existing loan token.
+   *
+   * @param {string} oldToken Previous tokens's symbol, id or creation tx (unique)
+   * @param {UpdateLoanToken} newTokenDetails
+   * @param {string} [newTokenDetails.symbol] New token's symbol (unique), no longer than 8
+   * @param {string} [newTokenDetails.name] Token's name, no longer than 128
+   * @param {string} [newTokenDetails.priceFeedId] Txid of oracle feeding the price
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:161]:
```suggestion
   * @param {string} [newTokenDetails.priceFeedId] token/currency pair to use for price of token
```

File: packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts (hunk 11 above)
Comment (surangap) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:141]: as I can see `should updateLoanToken by symbol as token` and `should updateLoanToken by given symbol` both does the same thing. i.e. `Update by symbol`. the only difference is the check/expect
Have a look again? if you think it's not duplicate, you can leave it.

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/updateLoanToken.test.ts:141]: Okay you are right. I thought you meant other thing

**Review by surangap — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]
